### DISCUSSION TITLE
JOIN query **much** slower on PostgREST than SQL editor

### DISCUSSION THREAD

**Discussion by diracdeltafunk** [supabase/supabase#16065] [Questions]
# The Problem

I have a REST query
```
Categories
?select=id,kb:KnowledgeBase!inner(proposition,value)
&kb.proposition=eq.c35a969b-8a67-4e49-85f8-52fe04c9cd9c
&kb.value=eq.true
```
which times out. However, the same query in the SQL editor in my browser completes very quickly!
```sql
explain analyze select c.id, kb.proposition, kb.value
from "Categories" c
join "KnowledgeBase" kb on kb.category = c.id
where
  kb.proposition = 'c35a969b-8a67-4e49-85f8-52fe04c9cd9c' and kb.value = true
;
```
yields
| QUERY PLAN                                                                                                                                     |
| ---------------------------------------------------------------------------------------------------------------------------------------------- |
| Nested Loop  (cost=0.42..1205.45 rows=13 width=33) (actual time=0.054..5.573 rows=8 loops=1)                                                   |
|   ->  Seq Scan on "KnowledgeBase" kb  (cost=0.00..1182.16 rows=13 width=33) (actual time=0.008..5.253 rows=8 loops=1)                          |
|         Filter: (value AND (proposition = 'c35a969b-8a67-4e49-85f8-52fe04c9cd9c'::uuid))                                                       |
|         Rows Removed by Filter: 53971                                                                                                          |
|   ->  Index Only Scan using "Categories_id_key" on "Categories" c  (cost=0.42..1.79 rows=1 width=16) (actual time=0.039..0.039 rows=1 loops=8) |
|         Index Cond: (id = kb.category)                                                                                                         |
|         Heap Fetches: 8                                                                                                                        |
| Planning Time: 1.875 ms                                                                                                                        |
| Execution Time: 5.636 ms                                                                                                                       |

*NB.* I am using the js library to make REST API requests.

# Steps Taken

1. **Turned RLS off.** This had no effect.
2. **Attempted `explain analyze` from js**

First, I set the timeout to 10 minutes. Then, I run
```js
let { data, error } = await supabase
    .from('Categories')
    .select('id,kb:KnowledgeBase!inner(proposition,value)')
    .match({ 'kb.proposition': 'c35a969b-8a67-4e49-85f8-52fe04c9cd9c', 'kb.value': true })
    .explain({ analyze: true })
```
After about 5 minutes, I get `data = null` and `error = { message: 'The upstream server is timing out' }`.

# Other Info

If I restrict the query further, I *can* get a response, which is why I say "**much** slower" instead of "hanging". For example:
```js
let { data, error } = await supabase
    .from('Categories')
    .select('id,kb:KnowledgeBase!inner(proposition,value)')
    .lte('morphisms', 5)
    .match({ 'kb.proposition': 'c35a969b-8a67-4e49-85f8-52fe04c9cd9c', 'kb.value': true })
    .explain({ analyze: true })
```
yields
```
Aggregate  (cost=506002.93..506002.95 rows=1 width=112) (actual time=1599.061..1599.063 rows=1 loops=1)
  ->  Limit  (cost=1317.41..505997.25 rows=379 width=48) (actual time=1554.087..1599.003 rows=6 loops=1)
        ->  Nested Loop  (cost=1317.41..505997.25 rows=379 width=48) (actual time=1554.085..1599.000 rows=6 loops=1)
              ->  Seq Scan on "Categories"  (cost=0.00..6687.49 rows=379 width=16) (actual time=0.076..27.033 rows=400 loops=1)
                    Filter: (morphisms <= '5'::smallint)
                    Rows Removed by Filter: 162681
              ->  Aggregate  (cost=1317.41..1317.42 rows=1 width=32) (actual time=3.929..3.929 rows=0 loops=400)
                    Filter: (json_agg("Categories_kb_1".*) IS NOT NULL)
                    Rows Removed by Filter: 1
                    ->  Subquery Scan on "Categories_kb_1"  (cost=0.00..1317.40 rows=1 width=41) (actual time=3.878..3.928 rows=0 loops=400)
                          ->  Limit  (cost=0.00..1317.39 rows=1 width=17) (actual time=3.878..3.927 rows=0 loops=400)
                                ->  Seq Scan on "KnowledgeBase" "KnowledgeBase_1"  (cost=0.00..1317.39 rows=1 width=17) (actual time=3.875..3.924 rows=0 loops=400)
                                      Filter: (value AND (proposition = 'c35a969b-8a67-4e49-85f8-52fe04c9cd9c'::uuid) AND (category = "Categories".id))
                                      Rows Removed by Filter: 53979
Planning Time: 1.815 ms
Execution Time: 1599.180 ms
```
(turning RLS off shaves about 20ms off of this time).

Compare this to the same query in the SQL editor:
```sql
explain analyze select c.id, kb.proposition, kb.value
from "Categories" c
join "KnowledgeBase" kb on kb.category = c.id
where
  c.morphisms <= 5
  and
  kb.proposition = 'c35a969b-8a67-4e49-85f8-52fe04c9cd9c' and kb.value = true
;
```
which yields
| QUERY PLAN                                                                                                                                |
| ----------------------------------------------------------------------------------------------------------------------------------------- |
| Nested Loop  (cost=0.42..1216.49 rows=1 width=33) (actual time=0.051..5.587 rows=6 loops=1)                                               |
|   ->  Seq Scan on "KnowledgeBase" kb  (cost=0.00..1182.16 rows=13 width=33) (actual time=0.008..5.274 rows=8 loops=1)                     |
|         Filter: (value AND (proposition = 'c35a969b-8a67-4e49-85f8-52fe04c9cd9c'::uuid))                                                  |
|         Rows Removed by Filter: 53971                                                                                                     |
|   ->  Index Scan using "Categories_id_key" on "Categories" c  (cost=0.42..2.64 rows=1 width=16) (actual time=0.038..0.038 rows=1 loops=8) |
|         Index Cond: (id = kb.category)                                                                                                    |
|         Filter: (morphisms <= 5)                                                                                                          |
|         Rows Removed by Filter: 0                                                                                                         |
| Planning Time: 1.991 ms                                                                                                                   |
| Execution Time: 5.667 ms                                                                                                                  |

---

Is there an obvious reason for the problem? If not, is there more information I could provide which would help troubleshoot the issue?

**Comment by diracdeltafunk** [comment]
Interestingly, if I perform a REST query with the join performed the other way around, the performance is much better!
```js
let { data, error } = await supabase
    .from('KnowledgeBase')
    .select('c:Categories!inner(id),proposition,value')
    .match({ 'proposition': 'c35a969b-8a67-4e49-85f8-52fe04c9cd9c', 'value': true })
    .explain({ analyze: true })
```
runs in about 6ms.
```
Aggregate  (cost=1216.94..1216.96 rows=1 width=112) (actual time=5.782..5.783 rows=1 loops=1)
  ->  Limit  (cost=0.42..1216.74 rows=13 width=49) (actual time=0.102..5.756 rows=8 loops=1)
        ->  Nested Loop  (cost=0.42..1216.74 rows=13 width=49) (actual time=0.101..5.754 rows=8 loops=1)
              ->  Seq Scan on "KnowledgeBase"  (cost=0.00..1182.16 rows=13 width=33) (actual time=0.008..5.345 rows=8 loops=1)
                    Filter: (value AND (proposition = 'c35a969b-8a67-4e49-85f8-52fe04c9cd9c'::uuid))
                    Rows Removed by Filter: 53971
              ->  Subquery Scan on "KnowledgeBase_c_1"  (cost=0.42..2.65 rows=1 width=40) (actual time=0.049..0.049 rows=1 loops=8)
                    ->  Limit  (cost=0.42..2.64 rows=1 width=16) (actual time=0.048..0.048 rows=1 loops=8)
                          ->  Index Only Scan using "Categories_id_key" on "Categories" "Categories_1"  (cost=0.42..2.64 rows=1 width=16) (actual time=0.047..0.047 rows=1 loops=8)
                                Index Cond: (id = "KnowledgeBase".category)
                                Heap Fetches: 8
Planning Time: 1.172 ms
Execution Time: 5.880 ms
```
However, this does not provide a solution for my use case. In general I need to perform queries like
```sql
select c.id, kb1.proposition, kb1.value, kb2.proposition, kb2.value, kb3.proposition, kb3.value
from "Categories" c
join "KnowledgeBase" kb1 on kb1.category = c.id
join "KnowledgeBase" kb2 on kb2.category = c.id
join "KnowledgeBase" kb3 on kb3.category = c.id
[...etc]
;
```
where I join on an arbitrary number of copies of `KnowledgeBase` (maybe 0).

In particular, if I need more than 1 copy of KnowledgeBase, this will not work (because there is no foreign key relation for the API to detect).

**Comment by diracdeltafunk** [comment]
@steve-chavez Sorry to bother you, but I've heard you're somewhat of an expert on the PostgREST implementation. Do you know how I could try to restructure my REST query to get its execution time closer to the SQL query?

**Reply by diracdeltafunk** [reply]
An alternate solution for me would be to rewrite my web server to connect to the postgres database directly, and then just use SQL. This should be fine, because I am not intending to ever run more than 2 instances of the server (and so there should not be more than ~2 active connections at any time)? But I am no expert, so if this seems like it might be a bad idea please let me know.

**Reply by steve-chavez** [reply]
@diracdeltafunk Thanks for the ping. How about wrapping your custom query in a function and calling it through [RPC](https://supabase.com/docs/reference/javascript/rpc)?

I'd also like to see what's wrong with the PostgREST generated query. Could you send me your project ref to steve@supabase.io?
